ocaml-lsp-server/src/ocaml_lsp_server.ml

```diff
@@ -553,140 +547,176 @@ let definition_query (state : State.t) uri position merlin_request =
+        let* doc = Fiber.return (Document_store.get store uri) in
+        let command =
+          Query_protocol.Occurrences (`Ident_at (Position.logical position))
+        in
+        let open Fiber.O in
+        let+ locs = Document.dispatch_exn doc command in
+        let loc =
+          List.find_opt locs ~f:(fun loc ->
+              let range = Range.of_loc loc in
+              Position.compare_inclusion position range = `Inside)
+        in
+        Ok (Option.map loc ~f:Range.of_loc))
+      ()
+  | Client_request.TextDocumentRename req -> later rename req
+  | Client_request.TextDocumentFoldingRange req -> later folding_range req
+  | Client_request.SignatureHelp req -> later signature_help req
+  | Client_request.ExecuteCommand _ -> not_supported
+  | Client_request.TextDocumentLinkResolve l -> now l
+  | Client_request.TextDocumentLink _ -> now None
+  | Client_request.WillSaveWaitUntilTextDocument _ -> now None
+  | Client_request.CodeAction params -> later code_action params
   | Client_request.CompletionItemResolve ci ->
-    let markdown =
-      markdown_support (client_capabilities state) ~field:(fun d ->
-          let open Option.O in
-          let+ completion = d.completion in
-          let* completion_item = completion.completionItem in
-          completion_item.documentationFormat)
-    in
-    let open Fiber.Result.O in
-    let resolve = Compl.Resolve.of_completion_item ci in
-    let* doc =
-      let uri = Compl.Resolve.uri resolve in
-      Fiber.return (Document_store.get state.store uri)
-    in
-    let* compl = Compl.resolve doc ci resolve query_doc ~markdown in
-    Fiber.return @@ Ok (compl, state)
+    later
+      (fun state () ->
+        let markdown =
+          markdown_support (client_capabilities state) ~field:(fun d ->
+              let open Option.O in
+              let+ completion = d.completion in
+              let* completion_item = completion.completionItem in
+              completion_item.documentationFormat)
+        in
+        let open Fiber.Result.O in
+        let resolve = Compl.Resolve.of_completion_item ci in
+        let* doc =
+          let uri = Compl.Resolve.uri resolve in
+          Fiber.return (Document_store.get state.store uri)
+        in
+        let* compl = Compl.resolve doc ci resolve query_doc ~markdown in
+        Fiber.return @@ Ok compl)
+      ()
   | Client_request.TextDocumentFormatting
       { textDocument = { uri }; options = _ } ->
-    let open Fiber.Result.O in
-    let uri = Uri.t_of_yojson (`String uri) in
-    let* doc = Fiber.return (Document_store.get store uri) in
-    Formatter.run rpc doc
-  | Client_request.TextDocumentOnTypeFormatting _ ->
-    Fiber.return @@ Ok (None, state)
-  | Client_request.SelectionRange req -> selection_range state req
+    later
+      (fun _ () ->
+        let open Fiber.Result.O in
+        let uri = Uri.t_of_yojson (`String uri) in
+        let* doc = Fiber.return (Document_store.get store uri) in
```

Comment:
Now that the requests are run concurrently, do we need to lock the access to the store?

Reply:
Currently, it doesn't impact correctness. The code that accesses the document store never yield. It would certainly be nice to add regardless, but I'm still thinking of the correct way to do it. Independent of this PR, we have lots of harmless imperfections of the form:

```
  let* doc = Fiber.return (Document_store.get store uri) in
  let client_capabilities = client_capabilities state in
  let open Fiber.O in
  let+ symbols = Document_symbol.run client_capabilities doc uri in
```

We are possibly using an out of date document by the time we acquire merlin. We need to delay getting the document until merlin is ready, rather than doing it eagerly. I'm not yet sure of a good api for this.